src/controller/SetUpCodePairer.cpp

```diff
@@ -0,0 +1,177 @@
+    return mBleLayer->NewBleConnectionByDiscriminator(discriminator, this, OnDiscoveredDeviceOverBleSuccess,
+                                                      OnDiscoveredDeviceOverBleError);
+#else
+    return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
+#endif // CONFIG_NETWORK_LAYER_BLE
+}
+
+CHIP_ERROR SetUpCodePairer::StopConnectOverBle()
+{
+#if CONFIG_NETWORK_LAYER_BLE
+    VerifyOrReturnError(mBleLayer != nullptr, CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE);
+    return mBleLayer->CancelBleIncompleteConnection();
+#else
+    return CHIP_NO_ERROR;
+#endif // CONFIG_NETWORK_LAYER_BLE
+}
+
+CHIP_ERROR SetUpCodePairer::StartDiscoverOverIP(uint16_t discriminator, bool isShort)
+{
+#if CHIP_DEVICE_CONFIG_ENABLE_MDNS
+    mCommissioner->RegisterDeviceDiscoveryDelegate(this);
+    Mdns::DiscoveryFilter filter(isShort ? Mdns::DiscoveryFilterType::kShort : Mdns::DiscoveryFilterType::kLong, discriminator);
+    return mCommissioner->DiscoverCommissionableNodes(filter);
+#else
+    return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
+#endif // CHIP_DEVICE_CONFIG_ENABLE_MDNS
+}
+
+CHIP_ERROR SetUpCodePairer::StopConnectOverIP()
+{
+#if CHIP_DEVICE_CONFIG_ENABLE_MDNS
+    mCommissioner->RegisterDeviceDiscoveryDelegate(nullptr);
+#endif // CHIP_DEVICE_CONFIG_ENABLE_MDNS
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR SetUpCodePairer::StartDiscoverOverSoftAP(uint16_t discriminator, bool isShort)
+{
+    return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
+}
+
+CHIP_ERROR SetUpCodePairer::StopConnectOverSoftAP()
+{
+    return CHIP_NO_ERROR;
+}
+
+void SetUpCodePairer::OnDeviceDiscovered(RendezvousParameters & params)
+{
+    LogErrorOnFailure(mCommissioner->PairDevice(mRemoteId, params.SetSetupPINCode(mSetUpPINCode)));
+}
+
+#if CONFIG_NETWORK_LAYER_BLE
+void SetUpCodePairer::OnDiscoveredDeviceOverBle(BLE_CONNECTION_OBJECT connObj)
+{
+    LogErrorOnFailure(StopConnectOverIP());
+    LogErrorOnFailure(StopConnectOverSoftAP());
+
+    Transport::PeerAddress peerAddress = Transport::PeerAddress::BLE();
+    RendezvousParameters params        = RendezvousParameters().SetPeerAddress(peerAddress).SetConnectionObject(connObj);
+    OnDeviceDiscovered(params);
+}
+
+void SetUpCodePairer::OnDiscoveredDeviceOverBleSuccess(void * appState, BLE_CONNECTION_OBJECT connObj)
+{
+    (static_cast<SetUpCodePairer *>(appState))->OnDiscoveredDeviceOverBle(connObj);
+}
+
+void SetUpCodePairer::OnDiscoveredDeviceOverBleError(void * appState, CHIP_ERROR err)
+{
+    LogErrorOnFailure(err);
+}
+#endif // CONFIG_NETWORK_LAYER_BLE
+
+#if CHIP_DEVICE_CONFIG_ENABLE_MDNS
+void SetUpCodePairer::OnDiscoveredDevice(const Mdns::DiscoveredNodeData & nodeData)
+{
+    LogErrorOnFailure(StopConnectOverBle());
+    LogErrorOnFailure(StopConnectOverIP());
+    LogErrorOnFailure(StopConnectOverSoftAP());
+
+    Transport::PeerAddress peerAddress = Transport::PeerAddress::UDP(nodeData.ipAddress[0], nodeData.port, nodeData.interfaceId[0]);
```

Comment:
Again, we should not be using interfaceId for anything other than link-local IPV6.

Reply:
Fixed.